### PR TITLE
ENH Adding skglm and gsroptim solvers

### PR DESCRIPTION
**Pull request by wassimmazouz**
- [x] Defining correct objective
- [x] Adding `skglm` solver
- [x] Working benchmark
- [x] Adding `gsroptim` solver 
- [x] `simple_config.yml` file

### CODE REVIEW

**Comment by wassimmazouz** [comment]
@mathurinm I'm not sure why the `make_correlated_data` dataset doesn't yield satisfying results. However, simulated data with `numpy` works quite well. [skglm solver benchmark](https://wassimmazouz.github.io/benchmarks/benchmark_sparse_group_lasso_skglm.html)

File: README.rst [hunk 1]
```diff
@@ -5,7 +5,7 @@ Benchmark repository for Sparse Group Lasso
 Benchopt is a package to simplify and make more transparent and reproducible the comparisons of optimization algorithms.
 The Sparse Group Lasso consists in solving the following program:
 
-$$ \\min_w \\, \\tfrac{1}{2} \\Vert y - Xw \\Vert^2_2 + \\lambda_1 \\Vert w \\Vert_1 + \\lambda_2 \\sum_g \\Vert w_g \\Vert_2 $$
+$$ \\min_w \\, \\tfrac{1}{2} \\Vert y - Xw \\Vert^2_2 + \\lambda_1 \\Vert w \\Vert_2 + \\lambda_2 \\sum_g \\Vert w_g \\Vert_1 $$
```
Comment (mathurinm) [inline README.rst:8]: It's the opposite (previous version was correct)

File: objective.py [hunk 2]
```diff
@@ -30,6 +37,27 @@ class Objective(BaseObjective):
...
 
+    def value_penalty(self, w):
+        """Value of penalty at vector ``w``."""
+        n_features = self.X.shape[1]
+        grp_indices, grp_ptr = grp_converter(self.grp_size, n_features)
+
+        n_groups = len(grp_ptr) - 1
+        alpha = 1
+
+        weights_groups = np.ones(n_groups, dtype=np.float64)
```
Comment (mathurinm) [inline objective.py:48]: this should not be hardcoded, it's the quivalent of alpha.
You're penalizing too much for make_correlated, hence you get a solution which is equal to 0 
see https://contrib.scikit-learn.org/skglm/tutorials/alpha_max.html#alpha-max :)

File: objective.py [hunk 3]
```diff
@@ -5,10 +5,13 @@
 # - getting requirements info when all dependencies are not installed.
 with safe_import_context() as import_ctx:
     import numpy as np
-
+    from skglm.utils.data import grp_converter
```
Comment (mathurinm) [inline objective.py:8]: skglm is a requirement of the benchmark, it should be stated

File: objective.py [hunk 4]
```diff
@@ -30,24 +37,50 @@ class Objective(BaseObjective):
...
         # returned by `Dataset.get_data`. This defines the benchmark's
         # API to pass data. This is customizable for each benchmark.
         # `set_data` can be used to preprocess the data.
         self.X, self.y = X, y
+        self.n_features = self.X.shape[1]
+        self.grp_indices, self.grp_ptr = grp_converter(self.grp_size,
+                                                       self.n_features)
+
+        self.n_groups = len(self.grp_ptr) - 1
+        alpha_max = norm(self.X.T @ y, ord=np.inf) / self.X.shape[0]
```
Comment (mathurinm) [inline objective.py:68]: this is not the correct alpha max for the sparse group lasso; the correct alpha_max is hard to compute but gsroptim has a function to compute it

File: objective.py [hunk 5]
```diff
@@ -30,24 +37,50 @@ class Objective(BaseObjective):
...
+        self.alpha = alpha_max / 1000
```
Comment (mathurinm) [inline objective.py:69]: you are using a single alpha, make it a parameter of the objective through `reg` as is done in the lasso case

File: solvers/gsroptim.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    from gsroptim.sgl import sgl_path
+
+
+class Solver(BaseSolver):
+    """Eugene Ndiaye, Olivier Fercoq, Alexandre Gramfort, and Joseph Salmon.
+    2016. GAP safe screening rules for Sparse-Group Lasso. In Proceedings
+    of the 30th International Conference on Neural Information Processing
+    Systems (NIPS'16). Curran Associates Inc., Red Hook, NY, USA, 388–396.
+    """
+    name = "gsrpotim"
```
Comment (mathurinm) [inline solvers/gsroptim.py:15]: typo

File: solvers/gsroptim.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    sampling_strategy = "iteration"
+
+    install_cmd = 'conda'
+    requirements = [
+        'pip:cython',
```
Comment (mathurinm) [inline solvers/gsroptim.py:20]: does cython need pip?

File: solvers/skglm.py [hunk 8]
```diff
@@ -20,32 +19,27 @@ class Solver(BaseSolver):
 
     install_cmd = 'conda'
     requirements = [
-        'pip:skglm'
+        "numpy'<2'",
+        "pip:git+https://github.com/scikit-learn-contrib/skglm.git#egg=skglm"
```
Comment (mathurinm) [inline solvers/skglm.py:23]: state why this is needed: sprase group is new in version 0.4 which is not released yet

**Comment by wassimmazouz** [comment]
@mathurinm the two solvers seem to converge to the same value. `benchopt run` is done in a few seconds. I used the `gsroptim.sgl.build_lambdas()` function to get the regularization parameters. I then divided by `2 * n_samples`, which, if I understood correctly, does the reverse work you did in this [PR](https://github.com/scikit-learn-contrib/skglm/issues/268).

File: objective.py [hunk 9]
```diff
@@ -18,36 +22,71 @@ class Objective(BaseObjective):
...
         # API to pass data. This is customizable for each benchmark.
         # `set_data` can be used to preprocess the data.
         self.X, self.y = X, y
+        self.n_samples, self.n_features = self.X.shape
+        self.grp_indices, self.grp_ptr = grp_converter(self.grp_size,
+                                                       self.n_features)
+        self.n_groups = len(self.grp_ptr) - 1
+
+        alpha_max = self.compute_alpha_max()
+        self.alpha = self.reg * alpha_max / 100
```
Comment (mathurinm) [inline objective.py:75]: why dividing alpha_max by 100 ?

File: objective.py [hunk 10]
```diff
@@ -63,5 +102,19 @@ def get_objective(self):
...
+            y=self.y,
+            alpha=self.alpha,
+            grp_size=self.grp_size,
+            grp_indices=self.grp_indices,
+            grp_ptr=self.grp_ptr
         )
+
+    def compute_alpha_max(self):
+        size_groups = self.grp_size * np.ones(
+            int(self.n_features / self.grp_size), order='F', dtype=np.intc)
```
Comment (mathurinm) [inline objective.py:114]: fortran order for 1 dimensional arrays should not change anything. Avoid anything that is not necessary

File: objective.py [hunk 11]
```diff
@@ -63,5 +102,19 @@ def get_objective(self):
...
-            y=self.y
+            y=self.y,
+            alpha=self.alpha,
+            grp_size=self.grp_size,
+            grp_indices=self.grp_indices,
+            grp_ptr=self.grp_ptr
         )
+
+    def compute_alpha_max(self):
+        size_groups = self.grp_size * np.ones(
```
Comment (mathurinm) [inline objective.py:113]: what's the difference between np.intc and np.int32 ?

File: objective.py [hunk 12]
```diff
@@ -63,5 +102,19 @@ def get_objective(self):
...
+            int(self.n_features / self.grp_size), order='F', dtype=np.intc)
+        omega = np.sqrt(size_groups)
```
Comment (mathurinm) [inline objective.py:115]: can you comment on this?

File: objective.py (hunk 10 above)
Comment (mathurinm) [inline objective.py:114]: this only supports contiguous groups, that have the same size. both skglm and gsroptim (to check for the latter) support variable size: add an error message if n_features is not a multiple of group_size, and add an issue in the repo to implement it later

File: objective.py [hunk 13]
```diff
@@ -63,5 +102,19 @@ def get_objective(self):
...
+        g_start = np.cumsum(size_groups, dtype=np.intc) - size_groups[0]
+
+        alpha_max = build_lambdas(self.X, self.y, omega, size_groups,
```
Comment (mathurinm) [inline objective.py:118]: there is a tau parameter equal to 1/2 in gsroptim, that makes the penalty have a scaling of 1/2 (check docstring of sgl_path).

Is this accounted for here?

**Comment by mathurinm** [comment]
> same

can you check that both solvers find 0 for alpha_max, something non zero for alpha_max / 1.05 ?
can you also check that they get the same solution on a decent range of lambdas, from lambda_max /1000 to lambda_max/ 2 ?

File: objective.py (hunk 11 above)
Comment (wassimmazouz) [reply objective.py:113]: `np.intc` might sometimes take 64 bits on some platforms, but it seems like most of the time it's equivalent to `np.int32`. It's used in the `gsroptim` [repo](https://github.com/EugeneNdiaye/Gap_Safe_Rules#:~:text=size_groups%20%3D%20size_group%20*%20np.ones(int(n_features%20/%20size_group)%2C%20order%3D%27F%27%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20dtype%3Dnp.intc)). I guess for our usage, `np.int32` is better.

**Comment by wassimmazouz** [comment]
@mathurinm could you take a look at this, I can't seem to find the right combinations of `alpha` and `lambda` to make this work. Based on this [test](https://github.com/scikit-learn-contrib/skglm/issues/268), `gsroptim` should take `lambda_max=self.alpha*2*n_samples` and `skglm` should take `self.alpha`. This is done in the code, but it doesn't give the same values of objective.

**Pull request merged**